test-workloads/src/bin/tellus.rs

```diff
@@ -81,6 +91,87 @@ fn set_vcpu_reg(vmid: u64, register: sbi::TvmCpuRegister, value: u64) {
     tsm::set_vcpu_reg(vmid, 0, register, value).expect("Tellus - TvmCpuGetRegister failed")
 }
 
+fn exercise_pmu_functionality() {
+    fn generate_counter_mask(high: u64) -> u64 {
+        (1 << high) - 1
+    }
+
+    /// Safety: Caller must ensure that that CSR exists on the platform, and that the read doesn't have unexpected side-effects
+    unsafe fn read_csr<const V: u16>() -> u64 {
+        let r: u64;
+        asm!("csrr {rd}, {csr}", rd = out(reg) r, csr = const V);
+        r
+    }
+
+    let num_counters = pmu::get_num_counters().expect("Tellus - GetNumCounters returned error");
+    for i in 0u64..num_counters {
+        let result = pmu::get_counter_info(i);
+        if result.is_err() {
+            println!("GetCounterInfo for {i} failed with {result:?}");
+        }
+    }
+
+    let event_type = PmuEventType::Hardware(PmuHardware::Instructions);
+    let config_flags = PmuCounterConfigFlags::default();
+    if let Ok(counter_index) = configure_matching_counters(
+        0,
+        generate_counter_mask(num_counters),
+        config_flags,
+        event_type,
+        0,
+    ) {
+        let start_flags = PmuCounterStartFlags::default();
+        let result = pmu::start_counters(counter_index, 0x1, start_flags, 0);
+        if result.is_err() {
+            println!("StartCounters failed with result {result:?}");
+        }
+
+        // Safety: CSR 0xC02 exists on the platform, and read has no side-effects
+        unsafe {
+            read_csr::<0xC02>();
+        }
+
+        let result = stop_counters(counter_index, 0x1, PmuCounterStopFlags::default());
+        if result.is_err() {
+            println!("StopCounters failed");
+        }
+
+        // Safety: CSR 0xC02 exists on the platform, and read has no side-effects
+        unsafe {
+            read_csr::<0xC02>();
+        }
+
+        let event_type = PmuEventType::Firmware(PmuFirmware::AccessLoad);
+        if let Ok(counter_index) = pmu::configure_matching_counters(
+            0,
+            generate_counter_mask(num_counters),
+            PmuCounterConfigFlags::default(),
+            event_type,
+            0,
+        ) {
+            if let Ok(value) = pmu::read_firmware_counter(counter_index) {
+                if value != 0 {
+                    println!("Firmware counter was 0 before start");
+                }
+
+                if pmu::start_counters(counter_index, 0x1, PmuCounterStartFlags::default(), 0)
+                    .is_ok()
+                {
+                    if pmu::read_firmware_counter(counter_index).is_err() {
+                        println!("Read firmware counter returned error");
+                    }
+                } else {
+                    println!("Start counter failed for FW counter");
+                }
+            }
+        } else {
+            println!("No matching FW counters were found");
+        }
```

Comment:
Don't we expect this to always be the case now?

Reply:
Yup...forgot to invert the condition.

Reply:
Changed.